Update dimensions to intuitive names + refactoring

bradyrx opened the pull request:
# Description

This PR will rename the dimensions expected by `climpred`. It's currently quite confusing, especially with `initialization` being a dimension name for reconstructions, observations, etc.

New dimensions:
`time`: any time series.
`lead`: forecast dimension of prediction ensemble; also dimension returned by prediction calculations.
`member`: ensemble member
`init`: initialization points, just for climate prediction ensemble.

Removed dimensions:
`initialization`: super confusing for a time series, especially when initializations don't occur every year.

New modules:
`metrics`: stores all metrics in one place
`comparisons`: stores all comparisons in one  place
`utils`: smaller functions, like `check_xarray` that are used in multiple places. will likely be updated to decorators in the future.

pytest:
* testing for reference ensembles and perfect model ensembles now working. 
* uses `open_dataset()` function for easy loading of our sample data.

Fixes https://github.com/bradyrx/climpred/issues/74, #100 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

All notebooks have been edited and now compile with the new dimension system. They show identical results to the previous notebooks. `pytest` scripts have been updated/fixed and run successfully.

# Checklist:

- [x] All notebooks run with this branch installed. (Notebooks can be updated to reflect changes)
- [x] Tests added for `pytest` if necessary. (updated `pytest` for perfect model so it will run. some issues need to be addressed to expand on this testing. also added tests for reference ensemble.)
- [x] `pytest` runs without breaking. (now this is the case)
- [x] I have added docstrings to all new functions.
- [x] I have updated the wiki with documentation, if necessary. (early efforts here: https://github.com/bradyrx/climpred/wiki/Dimension-Names)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have rebased onto master or develop (wherever I am merging) and dealt with any conflictions.
- [x] I have squashed commits to a reasonable amount.

# To Do
- [x] Finish updating dim names across all submodules (still need to do relative entropy)
- [x] Make sure that notebooks still compile
- [x] Finish refactoring metrics/comparisons
- [x] Move all helper functions to a submodule
- [x] See if it's reasonable to update `pytest` with this PR
- [x] Update sample data with new dimension names
- [x] Check/refactor private functions (https://www.datacamp.com/community/tutorials/role-underscore-python#name)

aaronspring commented:
[data types allowed]
`time`: any time series.   [int, datetime, cftime]
lead: forecast dimension of prediction ensemble; also dimension returned by prediction calculations. [int]
member: ensemble member [int (also names would be ok, but ints are easy)]

references would then have dimensions: time  (optional 2 spatial)
hindcasts would then have dimensions: time, lead, member (optional 2 spatial)

bradyrx commented:
> lead: forecast dimension of prediction ensemble; also dimension returned by prediction calculations. [int]. also very good that it is much shorter than initialization...

`initialization` is being replaced by `time`. `time` is being replaced by `lead`. But otherwise this is all correct.

> references/control would then have dimensions: time (optional 2 spatial)
hindcasts would then have dimensions: time, lead, member (optional 2 spatial)
here its important that hindcasts and reference/control have the same kind of time dim

Correct on this. `time` is standard time series for references/control. For the decadal prediction ensemble, it is actually the initialization years. But they should share that label to make calculations easier.

Another thought. Maybe the classes should check the data types of references/control/ensemble as it comes in.

aaronspring commented:
> Another thought. Maybe the classes should check the data types of references/control/ensemble as it comes in.

yep. but those are more high level ideas. lets first get the low level straight.

aaronspring commented:
I am still abit unsure about dropping `initialization`. `time` as a dimension for hindcast output can also be a bit misleading. more correct maybe would be `initializaiton time` (way to long). 

because hindcast.sel(time=1960) would basically provide the experiments started/initialized in 1960.

Maybe we just provide this as a longname or so in the metadata. I still think we should name it `time`. My comment is just to reduce confusion about what we mean.

bradyrx commented:
> I am still abit unsure about dropping initialization. time as a dimension for hindcast output can also be a bit misleading. more correct maybe would be initializaiton time (way to long).

I agree here. I like to avoid dimension names with spaces since you can't use the `ds.dim` command with spaces. Further, we are both away that "initialization" is so long. I don't think there's a perfect solution here. We use `S` in CESM for start date. I think the best option is clear documentation and appending meta data (long name) as you said.

Since we have to declare the dim to compare across for multiple computations here, we'd have to rename that dimension every time we ran that calculation. I.e., keeping it `time` just like in the reference objects/control ensures easy comparison. **Let me know your further thoughts about all this before we merge**.

aaronspring commented:
I still think `initialization` is the most correct word.
sorry for coming up with this concern so late. 

When writing the other comment, it appeared to me how odd it sounds to average over `time` (unclear) or `initializations`. It's a key term in predictions. `initialsation` is very close `start` or `S` the concept used to create the DPLE output files.

what about `init`? Still this needs to be documented, but the strength of `init` would be it's not as misleading as `time` (where I immeadiately get tricked into what we now will can `lead`).

I'm glad we agree on the data types for dims [comments above].

bradyrx commented:
> what about init? Still this needs to be documented, but the strength of init would be it's not as misleading as time (where I immeadiately get tricked into what we now will can lead).

Okay, I agree with you completely here. It's best we solve this now so that we don't have to think about it again as the package gets bigger. 

I think `init` is perfect. It's short and gets the point across, and we can document it more clearly in the wiki/Sphinx/tutorial, etc. 

I'll do some testing now, but I think we'll have to temporarily rename `init` to `time` when doing things like metrics. Since metrics need to know the dimension to compare over, we can't have one be `init` and then the control be `time`. Does that make sense?

I think we should differentiate the dimension namings and then just handle the dim mismatch under the hood so the user doesn't have to think about it.

aaronspring commented:
what about:

`short_name`|`types`|`long_name`
--- |---| ---
`lead`| `int`|`lead timestep after initialization [init]`
`init`|`int`,`pd.datetime`,`cftime`|`initialization: start date of experiment`
`member`|`int`,`str`|`ensemble member`

bradyrx commented:
The attribute assignment (`long_name`) can be done via something like https://github.com/bradyrx/climpred/issues/99

aaronspring commented:
> I'll do some testing now, but I think we'll have to temporarily rename init to time when doing things like metrics. Since metrics need to know the dimension to compare over, we can't have one be init and then the control be time. Does that make sense?

> I think we should differentiate the dimension namings and then just handle the dim mismatch under the hood so the user doesn't have to think about it.

Agreed: under the hood = inside the compute functions: rename control/reference to time to init or ds/hindcast to time.

aaronspring commented:
do the renaming, especially for the part you wrote. then I will look over the compute_perfect_model part. In the end you can have a look into the object stuff. Does that order make sense to you?

For me it does. Also I am clear on how the dummy testing date should look like.

bradyrx commented:
> do the renaming, especially for the part you wrote. then I will look over the compute_perfect_model part. In the end you can have a look into the object stuff. Does that order make sense to you?

Yep, it's seeming more straightforward than I first thought. I'll just request a review once I finish this for everything and you can look more carefully at the compute_perfect_model stuff.

bradyrx commented:
@aaronspring, I'm going to plan to merge this (eventually) without any changes to `relative_entropy`. I find that code very confusing right now since I haven't worked with it, and I cannot test if it works in the notebook, since I don't have access to the MPI files. 

I propose you open a small PR after this one to update dimension names in `relative_entropy` and maybe clean up some of the code/docstrings if it's straight forward.

bradyrx commented:
@aaronspring, okay this is ready for the most part if you want to look through. Probably only need to check `bootstrap` and `perfect_model` stuff closely. I've got all notebooks to compile with the same results as before (reconstruction, perfect model, and classes notebooks). As mentioned earlier, I'll have you open a PR just to fix `relative_entropy` and that notebook.

I'm going to do some light refactoring and will also check on the `pytest` stuff before merging this.

bradyrx commented:
Okay I have `pytest` working in a basic form right now for perfect model. Some issues I found while setting up `pytest`, but are out of the scope of this PR. (They weren't caught in the notebooks because they haven't been tried there. Also note these aren't bugs introduced by this PR, but have been around).

* `compute_persistence_pm` breaks when the control run doesn't have all of the init years in it https://github.com/bradyrx/climpred/issues/103
* `bootstrap_perfect_model` does not work for 'm2m' comparison
* `bootstrap_perfect_model` breaks with Datasets.

bradyrx commented:
Okay, this is totally ready to merge following review @aaronspring. I've got `pytest` working completely, squashed all my commits, updated notebooks, etc. I also refactored and introduced some new modules (`metrics`, `comparisons`, `utils`). Everything from this PR should make further development easier, since we have isolated submodules and a much clearer naming system for dimensions.

aaronspring commented:
>     * `compute_persistence_pm` breaks when the control run doesn't have all of the init years in it #103 

took out `_intersection` and used here.


>     * `bootstrap_perfect_model` does not work for 'm2m' comparison
 
...still checking why fails.

>     * `bootstrap_perfect_model` breaks with Datasets.

Thats because of the merging of results. This is not really the basic calculations of the bootstrapping function but some cosmetics which can be treated afterwards.

```python
    def _merge_result(result, new_result, new_result_name):
        new_result.name = new_result_name
        return xr.merge([result, new_result])
```
I exclude that here from the function. To be handled after the call of `bootstrap_perfect_model`. Now function returns 5 arguments.

Proposed way of handling the outputs into one file with one data_vars only. Therefore allowing for more variables to be calculated the same time:
```python
init_skill = compute_perfect_model(
    ds, control, metric='pearson_r', comparison='m2e')
pseudo_ds = pseudo_ens(ds, control)
uninit_skill = compute_perfect_model(
    pseudo_ds, control, metric='pearson_r', comparison='m2e')
pers_skill = compute_perfect_model(
    ds, control, metric='pearson_r', comparison='m2e')

skill = xr.concat([init_skill, uninit_skill, pers_skill], 'i')
skill['i'] = ['init', 'uninit', 'pers']

init_ci, uninit_ci, pers_ci, p_uninit_over_init, p_pers_over_init = res

p = xr.concat([i for i in [p_uninit_over_init, p_pers_over_init] if i is not None], 'i')
# what if one above is None?
p['i'] = ['uninit', 'pers']

ci = xr.concat([init_ci, uninit_ci, pers_ci], 'i').rename({'quantile':'results'})
ci['i'] = ['init', 'uninit', 'pers']

results = xr.concat([skill, p], 'results')
results['results']=['skill','p']
results = xr.concat([results,ci],'results')
results.coords
Coordinates:
  * i        (i) object 'init' 'pers' 'uninit'
    lon      (y, x) float64 161.9 163.6 165.2 166.9 ... -128.7 -127.1 -125.6
    lat      (y, x) float64 14.14 13.96 13.76 13.56 ... -32.44 -32.58 -32.72
  * lead     (lead) int64 1 2 3 4 5
  * results  (results) object 'skill' 'p' 0.025 0.975

```
This doesnt cover the case when we do not bootstrap something, eg. compute_persistence_skill=False, but thats again cosmetics.

With the _merge_results removed, also datasets pass.

>     * Because of the following line in `bootstrap_perfect_model`, it breaks when any PM metrics are used. This is because `compute_persistence_pm` has a flag in there to break if any non-xskillscore metrics are used.

switched off persistence then. And `compute_persistence_skill=False` in test.

bradyrx commented:
> EDIT: I will commit these changes soon for you to see the effect. 

Okay, all is good for me if it patches these issues and works for your workflow. Will you be committing them in a new PR? Should I merge this then you merge in the fixes? 

> I would change the testing data. We are faking TDD here anyways by first writing the production code and then the tests. If we constrain our tests to the specific MPI/CESM outputs we make ourselves even more dependent on those outputs. But I think it should work independent of that.

> I think it will be better to create sample dummy data (smaller) or just do the testing on 1D (we both agree that additional dimensions should just be kept.)

> I would get rid of the 3d stuff. we cannot provide proper sample_output. calc are anyways independent of extra variables. I would only do tests on 3d output where we really need 3d input as for rel_entropy.

I agree on all these comments. I haven't really done TDD so I didn't think about this, but it makes sense. I think I just used our sample data to have `pytest` not break. I'd like to merge this now (following your review/thumbs up), and then open a small PR to just change testing over to raw data that isn't dependent on MPI/CESM. Perhaps I would open this following your above comments, i.e., once we expect all types of bootstrapping cases to work.

aaronspring commented:
> then open a small PR to just change testing over to raw data that isn't dependent on MPI/CESM

agree. lets have real_data_tests (whihc might take some time) and synthetic_data_tests (which run very quickly)

aaronspring commented:
sorry for the mess I left with the two commits.

bradyrx commented:
Thanks for the updates. I'll clean up the Codacy issues and then merge. Enjoy your day off! I'll email you about setting up a skype call.

aaronspring commented:
It would be great if you could implement the new results for bootstrap the way I indicated. (message from the bus)

bradyrx commented:
I'll see if it makes sense and will give it a try.

bradyrx commented:
`pytest` and `treon` pass for this merge (minus the relative entropy notebook).

<img width="753" alt="Screen Shot 2019-04-30 at 2 48 23 PM" src="https://user-images.githubusercontent.com/8881170/56992334-04e5b300-6b57-11e9-8937-f27eccc1fc4b.png">

bradyrx commented:
@aaronspring, I had to remove the second commit you made. It broke `pytest`, the notebooks, etc. and wasn't very clear what was going on with all of the commenting. I've saved the changes you made in a separate branch that I'll open a PR for you to copy those edits over from.

> took out _intersection and used here.

This broke persistence_pm in many cases due to all of the checks on datetime after.

We can chat in our phonecall about this, but I propose you open a refactoring PR to finish up the perfect-model side of this stuff, and then we can build off of that. It could address:

* #103 (although maybe we should just not accept/expect situations where the control run doesn't contain all of the init years. could be an easy check).
* bootstrap_perfect_model does not work for 'm2m' comparison (this works on `pytest` for 1D now, but might be worth checking that 3D/maps work to confirm)
* bootstrap_perfect_model breaks with Datasets. (looks like you have a clear path here and started to work on it)
* Because of the following line in bootstrap_perfect_model, it breaks when any PM metrics are used. This is because compute_persistence_pm has a flag in there to break if any non-xskillscore metrics are used.  (I think this is fixed by just throwing an error when PM metrics are used + compute_persistence_skill=True).

Additional thoughts that are much needed:
* (priority) Update `relative_entropy` to have the new dimension names. I couldn't test this since I can't run the notebook
* See if you can make the relative entropy notebook work with dummy data so it can be tested with `treon`

For the phone call:
* I think it's important to discuss what the goal of `bootstrap_perfect_model` function is. I think having so many keywords and things returned would make it difficult for an outside user to manage.
